src/crimson/os/seastore/transaction_manager.h

```diff
@@ -335,65 +336,199 @@ class TransactionManager : public ExtentCallbackInterface {
-    // FIXME: existing_paddr can be absolute and pending
-    ceph_assert(existing_paddr.is_absolute());
-    assert(t.is_retired(existing_paddr, length));
-
-    SUBDEBUGT(seastore_tm, " laddr_hint: {} existing_paddr: {} length: {}",
-	      t, laddr_hint, existing_paddr, length);
-    auto bp = ceph::bufferptr(buffer::create_page_aligned(length));
-    bp.zero();
-
-    // ExtentPlacementManager::alloc_new_extent will make a new
-    // (relative/temp) paddr, so make extent directly
-    auto ext = CachedExtent::make_cached_extent_ref<T>(std::move(bp));
-
-    ext->init(CachedExtent::extent_state_t::EXIST_CLEAN,
-	      existing_paddr,
-	      PLACEMENT_HINT_NULL,
-	      NULL_GENERATION);
-
-    t.add_fresh_extent(ext);
-
-    return lba_manager->alloc_extent(
-      t,
-      laddr_hint,
-      length,
-      existing_paddr
-    ).si_then([ext=std::move(ext), laddr_hint, this](auto &&ref) {
-      ceph_assert(laddr_hint == ref->get_key());
-      ext->set_pin(std::move(ref));
-      return epm->read(
-        ext->get_paddr(),
-	ext->get_length(),
-	ext->get_bptr()
-      ).safe_then([ext=std::move(ext)] {
-	return map_existing_extent_iertr::make_ready_future<TCachedExtentRef<T>>
-	  (std::move(ext));
+    LBAPinRef &&pin,
+    extent_len_t remap_offset,
+    extent_len_t remap_len) {
+    LOG_PREFIX(TransactionManager::remap_pin);
+    auto remap_laddr = pin->get_key() + remap_offset;
+    auto remap_paddr = pin->get_val().add_offset(remap_offset);
+    // FIXME: paddr can be absolute and pending
+    ceph_assert(pin->get_val().is_absolute());
+    ceph_assert(remap_offset < pin->get_length());
+    ceph_assert(remap_offset + remap_len <= pin->get_length());
+
+    SUBDEBUGT(seastore_tm, "remapping on {} side,"
+      " original laddr: {}, original paddr: {}, original length: {}"
+      " remap laddr: {}, remap paddr: {}, remap length: {}", t,
+      remap_offset == 0 ? "left" : "right",
+      pin->get_key(), pin->get_val(), pin->get_length(),
+      remap_laddr, remap_paddr, remap_len);
+    return cache->get_extent_if_cached(
+      t, pin->get_val(), T::TYPE
+    ).si_then([this, &t, original_laddr = pin->get_key(),
+              remap_laddr, remap_paddr, remap_len](auto ext) {
+      ceph_assert(ext->is_clean());
+      std::optional<ceph::bufferptr> bp;
+      if (ext) {
+        ceph_assert(ext->get_length() > remap_len);
+        bp = ext->get_bptr();
+      }
+      return alloc_remapped_extent_with_retire<T>(
+        t,
+        remap_laddr,
+        remap_paddr,
+        remap_len,
+        original_laddr,
+        std::move(bp)
+      ).handle_error_interruptible(
+          remap_pin_iertr::pass_further{},
+          crimson::ct_error::assert_all{
+            "TransactionManager::remap_pin hit invalid error"
+          }
+      ).si_then([&t, remap_laddr](auto ext) {
+        ceph_assert(ext->get_laddr() == remap_laddr);
+        return remap_pin_iertr::make_ready_future<LBAPinRef>(
+          ext->get_pin().duplicate());
```

Comment:
This new interface is to remap the pin without reading the extent, this seems not supported after https://github.com/ceph/ceph/pull/47749.

I think there might be two solutions here:
1. Split `LBAManager::alloc_extent()` into two steps externally, first to `LBAManager::alloc_mapping()`, and then to link to the extent. So this `remap_pin()` can reuse `alloc_mapping()` to get the intermediate pin to return.
2. Implement and reintroduce `LogicalCachedExtent::get_pin()`, seems also possible through `ChildableCachedExtent::parent_tracker`.

Does it also make sense to you ? @xxhdx1985126

Reply:
Maybe I implement the first solution first?

Reply:
Reasonable to me.

Reply:
> This new interface is to remap the pin without reading the extent, this seems not supported after #47749.
> 
> I think there might be two solutions here:
> 
> 1. Split `LBAManager::alloc_extent()` into two steps externally, first to `LBAManager::alloc_mapping()`, and then to link to the extent. So this `remap_pin()` can reuse `alloc_mapping()` to get the intermediate pin to return.
> 2. Implement and reintroduce `LogicalCachedExtent::get_pin()`, seems also possible through `ChildableCachedExtent::parent_tracker`.
> 
> Does it also make sense to you ? @xxhdx1985126

I think it's fine. BTW, when developping cloning in seastore, I also encountered an issue about splitting extents, so I implemented a ["LBAManager::split_mapping()" interface](https://github.com/ceph/ceph/pull/51141/commits/73c17d28c527515e1e33068abb135349253c6482), which I think may also benefit you.

Reply:
> so I implemented a "LBAManager::split_mapping()" interface

Yeah, that looks to be a referencable implementation. The requrement in solution 1 is a sub-step of `LBAManager::split_mapping()`, which only needs the remapping part.

Reply:
> I think it's fine. BTW, when developping cloning in seastore, I also encountered an issue about splitting extents, so I implemented a ["LBAManager::split_mapping()" interface](https://github.com/ceph/ceph/pull/51141/commits/73c17d28c527515e1e33068abb135349253c6482), which I think may also benefit you.

ok, thanks!